connect/src/com/telenor/mobileconnect/MobileConnectSdkProfile.java

```diff
@@ -0,0 +1,317 @@
+    }
+
+    @Override
+    public String getClientSecret() {
+        return operatorDiscoveryResult.getClientSecret();
+    }
+
+    @Override
+    public String getRedirectUri() {
+        return operatorDiscoveryConfig.getOperatorDiscoveryRedirectUri();
+    }
+
+    @Override
+    public String getExpectedIssuer(String actualIssuer) {
+        // discrepancy between .well-known configuration and the actual issuer returned
+        if (operatorDiscoveryResult.getBasePath().contains("telenordigital.com")) {
+            return actualIssuer;
+        }
+        if (getWellKnownConfig() != null) {
+            return getWellKnownConfig().getIssuer();
+        }
+        return actualIssuer;
+    }
+
+    @Override
+    public List<String> getExpectedAudiences(List<String> actualAudiences) {
+        return new ArrayList<>(actualAudiences);
+    }
+
+    @Override
+    public Uri getAuthorizeUri(Map<String, String> parameters, List<String> locales) {
+        Uri.Builder builder = ConnectUrlHelper.getAuthorizeUriStem(
+                parameters,
+                getClientId(),
+                getRedirectUri(),
+                locales,
+                getApiUrl())
+                .buildUpon();
+        for (String pathSeg : getApiUrl().pathSegments()) {
+            builder.appendPath(pathSeg);
+        }
+        return builder.build();
+    }
+
+    @Override
+    public DoNext onStartAuthorization(Map<String, String> parameters) {
+        OperatorDiscoveryAPI.OperatorDiscoveryResult odResult = null;
+
+        final String msisdn = readPhoneNumber();
+        if (msisdn != null) {
+            odResult = getOperatorDiscoveryResult(msisdn);
+        }
+
+        if (odResult != null) {
+            parameters.put("login_hint", String.format("ENCR_MSISDN:%s", odResult.getSubscriberId()));
+        }
+
+        if (isInitialized || initialize(odResult)) {
+            return DoNext.proceed;
+        }
+
+        return DoNext.cancel;
+    }
+
+    private String readPhoneNumber() {
+        TelephonyManager phMgr = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+        try {
+            String msisdn = phMgr.getLine1Number();
+            return (TextUtils.isEmpty(msisdn) ? null : msisdn);
+        } catch (RuntimeException ignored) {
+        }
+        return null;
+    }
+
+    private OperatorDiscoveryAPI.OperatorDiscoveryResult getOperatorDiscoveryResult() {
+        TelephonyManager phMgr = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
+        String networkOperator = phMgr.getNetworkOperator();
+        if (!TextUtils.isEmpty(networkOperator)) {
+            final String mcc = networkOperator.substring(0, 3);
+            final String mnc = networkOperator.substring(3);
+            Future<OperatorDiscoveryAPI.OperatorDiscoveryResult> operatorDiscoveryResultFuture =
+                    sExecutor.submit(new Callable<OperatorDiscoveryAPI.OperatorDiscoveryResult>() {
```

Comment:
I'm not sure I understand this peace of code.
On the surface it looks like it is asynchronous, but actually it's quite synchronous isn't it?
Then why do we need the ExecutorService, and the Future if we could just make a synchronous retrofit call?
Also if you decide to refactor it to asynchronous, maybe you should use the facilities provided retrofit for making asynchronous calls, instead of implementing ones yourself.
I think a good example is in ConnectId SDK's part: getAccessTokenFromCode.